src/cmd/input.go

```diff
@@ -30,34 +34,108 @@ func readInputConfig() {
 	viper.ReadInConfig()
 }
 
-func readResourceInput[T any]() (*T, error) {
+func readInput() ([]byte, error) {
 	var err error
-	var resource T
-	var yamlData []byte
+	var input []byte
 
 	switch dataFile {
 	case ".":
-		yamlData, err = os.ReadFile("./data.yaml")
+		input, err = os.ReadFile("./data.yaml")
 	case "-":
 		if isStdInFromTerminal() {
 			log.Info().Msg("Reading input directly from command line... Press CTRL+D to stop typing")
 		}
 		buf := bytes.Buffer{}
 		_, err = buf.ReadFrom(os.Stdin)
-		yamlData = buf.Bytes()
+		input = buf.Bytes()
 	default:
-		yamlData, err = os.ReadFile(dataFile)
+		input, err = os.ReadFile(dataFile)
 	}
 	if err != nil {
-		return nil, err
+		return input, err
 	}
+	return input, nil
+}
 
-	if err := yaml.Unmarshal(yamlData, &resource); err != nil {
+func ReadResource[T any](input []byte) (*T, error) {
+	var resource T
+	if err := yaml.Unmarshal(input, &resource); err != nil {
 		return nil, err
 	}
 	return &resource, nil
 }
 
+func handleJSONSchema(m map[string]any) error {
+	// since a YAML object is map[string]any which is compatible with JSONSchema so nothing needs to be done.
+	// if schema is a string, the JSON inside the string needs to be parsed. supports multiline.
+	if _, ok := m["schema"].(string); !ok {
+		return nil
+	}
+	jsonSchema, err := opslevel.NewJSONSchema(m["schema"].(string))
+	if err != nil {
+		return err
+	}
+	m["schema"] = jsonSchema
+	return nil
+}
+
+func handleJSONString(m map[string]any) error {
+	// TODO: multiline JSON is not supported as a value input because handling whitespace and escape characters is nontrivial
+	jsonString, err := opslevel.NewJSONInput(m["value"])
+	if err != nil {
+		return err
+	}
+	m["value"] = jsonString
+	return nil
+}
+
+func ReadResourceHandleJSONFields[T any](input []byte) (*T, error) {
+	var err error
+	if input == nil {
+		input, err = readInput()
+		if err != nil {
+			return nil, fmt.Errorf("error reading from input: %w", err)
+		}
+	}
+
+	m, err := ReadResource[map[string]any](input)
+	if err != nil {
+		return nil, fmt.Errorf("error creating map from input: %w", err)
+	}
+	toMap := *m
+
+	if _, ok := toMap["schema"]; ok {
```

Comment:
this feels wrong.  This feels like we need to implement custom yaml mashaller functions for `PropertyInput` and `PropertyDefinitionInput` ?

Reply:
@rocktavious will address that in the follow-up PR that handles `JSONSchema`/`JSONString`